### PR TITLE
Add disjoint sets data structure

### PR DESCRIPTION
**Pull request by haozhun**

### CODE REVIEW

File: presto-main/src/main/java/com/facebook/presto/util/DisjointSet.java [hunk 1]
```diff
@@ -0,0 +1,148 @@
...
+        }
+
+        public Entry(Entry<T> entry)
+        {
+            this(entry.parent, entry.rank);
+        }
+
+        private Entry(T parent, int rank)
+        {
+            this.parent = parent;
```
Comment (erichwang) [inline presto-main/src/main/java/com/facebook/presto/util/DisjointSet.java]: add some checks here for rank being >= 0

File: presto-main/src/main/java/com/facebook/presto/util/DisjointSet.java [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+        }
+
+        public void incrementRank()
+        {
+            checkState(parent == null);
+            rank++;
+        }
+    }
+
+    private final HashMap<T, Entry<T>> map;
```
Comment (erichwang) [inline presto-main/src/main/java/com/facebook/presto/util/DisjointSet.java]: HashMap => Map

File: presto-main/src/main/java/com/facebook/presto/util/DisjointSet.java [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+
+    public DisjointSet()
+    {
+        map = new HashMap<>();
```
Comment (erichwang) [inline presto-main/src/main/java/com/facebook/presto/util/DisjointSet.java]: you can just put this in the field and not need a constructor

File: presto-main/src/main/java/com/facebook/presto/util/DisjointSet.java [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import static com.google.common.base.Preconditions.checkState;
+
+public class DisjointSet<T>
```
Comment (erichwang) [inline presto-main/src/main/java/com/facebook/presto/util/DisjointSet.java]: would be nice to include 2 convenience methods:
1) something that maps from every T to Set<T>
2) something to test if a T exists in the DisjointSet without adding it

**Comment by erichwang** [comment]
otherwise looks good

**Comment by haozhun** [comment]
I addressed first two comments. For the additional convenience method, feel free to add it.

**Pull request merged**